mlx_lm/awq.py

```diff
@@ -0,0 +1,541 @@
+
+            w_q = quantize_func(w_m)
+
+            w_q = w_q.reshape(*w_q.shape[:-1], -1, group_size)
+            out_q = mx.einsum("btdg,odg->btod", x, w_q)
+
+            # Take the mean across the input batch
+            loss = mse(out, out_q).sum(axis=(0, 1))
+            if group is not None:
+                loss = mx.distributed.all_sum(loss, stream=mx.cpu) / group.size()
+            loss /= out.shape[0] * out.shape[1]
+            best_indices = loss < best_error
+            best_error = mx.where(best_indices, loss, best_error)
+            best_w_max = mx.where(best_indices[..., mx.newaxis], w_max, best_w_max)
+            best_w_min = mx.where(best_indices[..., mx.newaxis], w_min, best_w_min)
+            mx.eval(best_w_max, best_w_min, best_error)
+
+        w_max_all.append(best_w_max)
+        w_min_all.append(best_w_min)
+
+    best_w_max = mx.concatenate(w_max_all, axis=0)
+    best_w_min = mx.concatenate(w_min_all, axis=0)
+
+    w_r = w_all.reshape(*w_all.shape[:-1], -1, group_size)
+    best_w = mx.clip(w_r, best_w_min, best_w_max)
+    best_w = best_w.reshape(w_all.shape)
+
+    mx.eval(best_w)
+    return best_w
+
+
+def clip_block(
+    block: nn.Module, quantize_func: Callable, group_size: int, n_grid: int = 20
+):
+
+    def apply_clip(path, module):
+        if (
+            isinstance(module, nn.Linear)
+            and "q_proj" not in path
+            and "k_proj" not in path
+        ):
+            best_weight = search_best_clip(
+                module,
+                quantize_func=quantize_func,
+                group_size=group_size,
+                n_grid=n_grid,
+            )
+            module.weight = best_weight
+
+    tree_map_with_path(apply_clip, block.leaf_modules(), is_leaf=nn.Module.is_module)
+
+
+def awq_quantize(
+    model,
+    inputs: mx.array,
+    group_size: int = 64,
+    bits: int = 3,
+    embed_group_size: int = 32,
+    embed_bits: int = 4,
+    n_grid: int = 20,
+):
+    group = mx.distributed.init()
+
+    def quantize_func(w):
+        wq = mx.quantize(w, bits=bits, group_size=group_size)
+        return mx.dequantize(*wq, bits=bits, group_size=group_size)
+
+    mask = create_attention_mask(inputs)
+
+    model.model.embed_tokens = model.model.embed_tokens.to_quantized(
+        group_size=embed_group_size, bits=embed_bits
+    )
+    inputs = model.model.embed_tokens(inputs)
+
+    def capture(module):
+        if not isinstance(module, (nn.Linear, SwitchLinear)):
+            return module
+
+        class Catcher(nn.Module):
+            def __call__(self, x: mx.array, *args, **kwargs):
+                # Store the input features on the original modules.
+                if hasattr(module, "input_feat"):
+                    module.input_feat = mx.concatenate([module.input_feat, x], axis=0)
+                else:
+                    module.input_feat = x
+
+                # Also store the MOE indices if applicabale
+                module.indices = None
```

Comment:
Does it make sense to always set this to `None` here? I might be misunderstanding.. but it seems like you won't be able to aggregate the indices across multiple calls if they are always set to `None`

Reply:
Fixed.